File: DAA/3-fractKnapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Item {
    int weight;
    int value;
    double valuePerWeight;

    Item(int w, int v) : weight(w), value(v) {
        valuePerWeight = static_cast<double>(value) / weight;
    }
};

bool compareItems(const Item& item1, const Item& item2) {
    return item1.valuePerWeight > item2.valuePerWeight;
}

double fractionalKnapsack(int W, vector<Item>& items) {
    sort(items.begin(), items.end(), compareItems);

    double totalValue = 0.0;
    int currentWeight = 0;

    for (const Item& item : items) {
        if (currentWeight + item.weight <= W) {
            totalValue += item.value;
            currentWeight += item.weight;
        } else {
            int remainingWeight = W - currentWeight;
            totalValue += item.valuePerWeight * remainingWeight;
            break;
        }
    }

    return totalValue;
}

int main() {
    int W = 50;
    vector<Item> items = {Item(10, 60), Item(20, 100), Item(30, 120)};

    double maxValue = fractionalKnapsack(W, items);
    cout << "Maximum value: " << maxValue << endl;

    return 0;
}

File: DAA/3-fractKnapsack.py
class Item:
    def __init__(self, weight, profit):
        self.weight = weight
        self.profit = profit
        self.profit_per_weight = profit / weight

def fractional_knapsack(W, items):
    items.sort(key=lambda x: x.profit_per_weight, reverse=True)

    total_profit = 0.0
    current_weight = 0

    for item in items:
        if current_weight + item.weight <= W:
            total_profit += item.profit
            current_weight += item.weight
        else:
            remaining_weight = W - current_weight
            total_profit += item.profit_per_weight * remaining_weight
            break

    return total_profit

if __name__ == "__main__":
    W = 15
    items = [Item(2,10), Item(3,5), Item(5,15), Item(7,7), Item(1,6), Item(4,18), Item(1,3)]

    max_profit = fractional_knapsack(W, items)
    print("Maximum profit:", max_profit)
    
